pilopt/src/lib.rs

```diff
@@ -31,6 +34,48 @@ pub fn optimize<T: FieldElement>(mut pil_file: Analyzed<T>) -> Analyzed<T> {
     pil_file
 }
 
+/// Just perform some very light optimizations (inlining constants and evaluating
+/// constant expressions).
+pub fn optimize_mildly<T: FieldElement>(mut pil_file: Analyzed<T>) -> Analyzed<T> {
+    inline_constant_values(&mut pil_file);
+    evaluate_constant_subtrees(&mut pil_file);
+    pil_file
+}
+
+/// Inlines references to symbols with a single constant value.
+fn inline_constant_values<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    let constants = &pil_file.constants.clone();
+    pil_file.post_visit_expressions_mut(&mut |e| match e {
+        Expression::Reference(Reference::Poly(poly)) => {
+            if !poly.next && poly.index.is_none() {
+                if let Some(value) = constants.get(&poly.name) {
+                    *e = Expression::Number(*value)
+                }
+            }
+        }
+        Expression::Constant(name) => *e = Expression::Number(constants[name]),
+        _ => {}
+    });
+}
+
+/// Substitutes expression that evaluate to a constant value.
+fn evaluate_constant_subtrees<T: FieldElement>(pil_file: &mut Analyzed<T>) {
```

Comment:
Same comment as [here](https://github.com/powdr-labs/powdr/pull/647/files#r1342853191)

Reply:
I tried to use `evaluate_expression`, but it's way too heavy machinery for this here.

Since it uses `post_visit`, it will first check the subtrees. If they both evaluate to a number, it will also replace the current node by a number. So it works for trees of arbitrary depth.

Reply:
Ah cool, nice!